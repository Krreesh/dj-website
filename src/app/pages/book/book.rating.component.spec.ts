import {ComponentFixture, TestBed} from '@angular/core/testing';
import {BookRatingComponent} from './book.rating.component';

describe('BookRatingComponent', () => {
  let component: BookRatingComponent;
  let fixture: ComponentFixture<BookRatingComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [BookRatingComponent]
    });
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(BookRatingComponent);
    component = fixture.componentInstance;
  });

  describe('when the rating has no remainder', () => {
    it('shows five empty stars when the rating is zero', () => {
      component.rating = 0;
      fixture.detectChanges();

      expect(getEmptyStars().length).toEqual(5);
    });

    it('shows zero empty stars when the rating is five', () => {
      component.rating = 5;
      fixture.detectChanges();

      expect(getEmptyStars().length).toEqual(0);
    });

    it('shows "five minus rating" empty stars', () => {
      component.rating = 3;

      fixture.detectChanges();

      expect(getEmptyStars().length).toEqual(2);
    });

    it('shows five full stars when the rating is five', () => {
      component.rating = 5;
      fixture.detectChanges();

      expect(getFullStars().length).toEqual(5);
    });

    it('shows zero full stars when the rating is zero', () => {
      component.rating = 0;
      fixture.detectChanges();

      expect(getFullStars().length).toEqual(0);
    });

    it('shows "rating" full stars', () => {
      component.rating = 3;
      fixture.detectChanges();

      expect(getFullStars().length).toEqual(3);
    });

    it('shows no half star', () => {
      component.rating = 3;
      fixture.detectChanges();

      expect(getHalfStar()).toBeNull();
    });
  });

  describe('when the rating has a remainder', () => {
    const ratingWithRemainder = 2.5;

    it('shows a half star', () => {
      component.rating = ratingWithRemainder;

      fixture.detectChanges();

      expect(getHalfStar()).not.toBeNull();
    });

    it('shows 5 - (ratingWithRemainder + 0.5) empty stars', () => {
      component.rating = ratingWithRemainder;

      fixture.detectChanges();

      expect(getEmptyStars().length).toEqual(2);
    });

    it('shows ratingWithRemainder - 0.5 full stars', () => {
      component.rating = ratingWithRemainder;
      fixture.detectChanges();

      expect(getFullStars().length).toEqual(2);
    });
  });

  function getEmptyStars(): HTMLSpanElement[] {
    return fixture.nativeElement.querySelectorAll('[data-jest="empty star"]');
  }

  function getFullStars(): HTMLSpanElement[] {
    return fixture.nativeElement.querySelectorAll('[data-jest="full star"]');
  }

  function getHalfStar(): HTMLSpanElement {
    return fixture.nativeElement.querySelector('[data-jest="half star"]');
  }
});
