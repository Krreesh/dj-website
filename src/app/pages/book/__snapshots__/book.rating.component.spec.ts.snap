// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookRatingComponent on init shows the stars in the right order, when the rating is in between 0 and 5 1`] = `
<div
  class="book-rating"
  data-jest="stars"
>
  
  <span
    class="fa fa-star checked"
    data-jest="full star"
  />
  <span
    class="fa fa-star checked"
    data-jest="full star"
  />
  <span
    class="fa fa-star checked"
    data-jest="full star"
  />
  
  <span
    class="fa fa-star-half-empty"
    data-jest="half star"
  />
  
  <span
    class="fa fa-star-o"
    data-jest="empty star"
  />
</div>
`;
